### DISCUSSION TITLE
Problem with long exposures

### DISCUSSION THREAD

**Discussion by lboclboc** [lboclboc/indi-3rdparty#7] [General]
> From: QRainman commented 6 hours ago
> I created another pull request for the color fixes which also fixed the problem with the cut off pixels at the right side of the frame. 
> I'm not really happy that the color pattern is exactly the opposite of the documentation but i also can't find out where there that
> +/- 1 offset is introduced, so i'm leaving i suggest to leave it like that for now.

> Does your master already contain the fix for the exposure time? My cameras still don't go beyond 1s for V1 and 6s for V2.

I believed that my master should have fixed the long exposure after i moved the enabling of the encoder to just before capture. But I think I have seen once that it did not work still (even after a waste frame). So, its not certain.

**Comment by QRainman** [comment]
So i think i figured out how to force it into long exposure mode (at least for the V1 camera for now).
```
 MMAL_STATUS_T status = mmal_port_parameter_set_uint32(component->control, MMAL_PARAMETER_CAMERA_CUSTOM_SENSOR_CONFIG, 3);
```
To work reliably, this needs to be called before the camera component is enabled. On the V1 camera this allows settings form 0.5s up to 6s then. The enable is currently in the constructor and if it has to run before it means it can't be changed anymore without disconnecting/reconnecting. There is this comment before the enable:
```
// FIXME: moved from #001 to after sensor Enable the controlport so calls below works.
```
So i'm guessing it's there for a good reason. Any suggestions how to consolidate those two things?

**Reply by lboclboc** [reply]
Hi, This is a left over comment from my desperate testing to make it work. If I remember correctly you need to enable the camera quite early on, especially to make the workaround work to have one waste frame before the proper exposure kicks in. 
If this makes it totally predicable I guess everything can be moved to the start of capture part.

**Comment by lboclboc** [comment]
Ok, done some testing with the HiQ-camera now with both forcing the mode (before enabling the camera), also tested some with burst-mode which I got a suggestion from the forum about. But, no change what I could see.
I did update my fork so it will write the actual exposure time into the FITS-image, at least the frames with bad exposure should be easy to sort out then. 
I also set the sensor config mode for all cameras for the correct frame rate. For the HiQ, format 3 should always be used.
When the raspistill is used for long exposure it takes very long time, I think it is actually taking some internal waste frames to get the camera to synchronize (haven't had time to examine their code any more regarding this).

**Reply by QRainman** [reply]
I think those additional frames are for the auto white balance, even if it's switched off. For the old cameras i never could get it to take less than two frames...

> I did update my fork so it will write the actual exposure time into the FITS-image

Thank you! I was already pulling what's left of my hair out, trying to figure out how to do that...

Actually wanted to let you know that it looks like the sensor mode has to be set before the exposure mode is set to 'off'. I had a test run with my V1 camera and it did not work unless i moved the sensor mode setting before that. Just to be sure, i moved it now all the way up to the top of `MMALCamera::setExposureParameters`. The first image i take is still restricted to <= 1s but at least all the following images seem to be working fine from then on. Works even if i go below 1 second later and back up again.
